### PR TITLE
Support SQL array subscript operator

### LINKED ISSUE #3584: Support `NOT IN`
Currently we do not support `NOT IN` queries:

```sql
select * from a where x not in ('a', 'b', 'c');
```

This should be supported.

### PR DESCRIPTION
**Pull request by hatyo**
This introduces support to SQL array subscription operator, with the following caveats:
- It works with one-dimensional arrays only.
- Similar to other vendors, such as Postgres and Oracle, it returns `null` when the index is out-of-bound instead of raising an error.
- Uses 1-based index, as defined in SQL standard.

Example:

```sql
create table A(pk integer, x integer array, primary key(pk))
insert into A values (1, [1, 2, 3]), (2, [2, 3, 4]), (3, [3, 4, 5])
select X[2] from A where pk = 1;
-- returns 2
select X[1+1] from A where pk = 1;
-- returns 2
select X[-1] from A where pk = 1;
-- returns null
select X[1000] from A where pk = 1;
-- returns null
```

This also restructures the parsing rules for expressions such that it prohibits more invalid combinations of binary expressions, and aligns them to work well with the subscript operator, and adds support to `not in` which was previously throwing a syntax error.

This fixes #3584, #3570, and partially #3588 with few limitations.

### CODE REVIEW

**Comment by arnaud-lacurie** [comment]
Could you call out in the description that the error code has changed for a certain operation? So that we can surface it in our release notes.

File: yaml-tests/src/test/resources/arrays.yamsql [hunk 1]
```diff
@@ -65,4 +65,35 @@ test_block:
...
+    -
+      # out of bound
+      - query: select X[100000] AS ERR FROM A WHERE pk = 1
+      - supported_version: !current_version
+      - result: [{ERR: !null _}]
+    -
+      # out of bound
+      - query: select X[-1] AS ERR FROM A WHERE pk = 1
+      - supported_version: !current_version
+      - result: [{ERR: !null _}]
```
Comment (arnaud-lacurie) [inline yaml-tests/src/test/resources/arrays.yamsql:103]: Would creating an index on `X[1]` be supported out of the box? If so, could you add a test for it?

File: yaml-tests/src/test/resources/arrays.yamsql [hunk 2]
```diff
@@ -65,4 +65,35 @@ test_block:
...
+    -
+      # high-order subscript indexes are supported
+      - query: select pk FROM A WHERE X[X[1] - X[1] + 1] between 0 + X[1+0] and X[1-0] - 0
+      - supported_version: !current_version
+      - result: [{PK: 1}, {PK: 2}, {PK: 3}]
+    -
+      # out of bound
+      - query: select X[100000] AS ERR FROM A WHERE pk = 1
+      - supported_version: !current_version
+      - result: [{ERR: !null _}]
```
Comment (arnaud-lacurie) [inline yaml-tests/src/test/resources/arrays.yamsql:98]: Could you add a test with `X[0]`?
And another test where the array is empty?

File: fdb-relational-core/src/main/java/com/apple/foundationdb/relational/recordlayer/query/functions/SqlFunctionCatalogImpl.java [hunk 3]
```diff
@@ -122,6 +122,7 @@ private static ImmutableMap<String, Function<Integer, Optional<BuiltInFunction<?
                 .put("&", argumentsCount -> BuiltInFunctionCatalog.resolve("bitand", argumentsCount))
                 .put("|", argumentsCount -> BuiltInFunctionCatalog.resolve("bitor", argumentsCount))
                 .put("^", argumentsCount -> BuiltInFunctionCatalog.resolve("bitxor", argumentsCount))
+                .put("[", argumentsCount -> BuiltInFunctionCatalog.resolve("subscript", argumentsCount))
```
Comment (arnaud-lacurie) [inline fdb-relational-core/src/main/java/com/apple/foundationdb/relational/recordlayer/query/functions/SqlFunctionCatalogImpl.java]: Could the key be `[]` instead of `[`?

File: fdb-relational-core/src/main/antlr/RelationalParser.g4 [hunk 4]
```diff
@@ -1129,43 +1129,38 @@ namedFunctionArg
...
+    | fullColumnName                                                                      #fullColumnNameExpressionAtom // done
+    | functionCall                                                                        #functionCallExpressionAtom // done
+    | preparedStatementParameter                                                          #preparedStatementParameterAtom // done
+    | recordConstructor                                                                   #recordConstructorExpressionAtom // done
+    | arrayConstructor                                                                    #arrayConstructorExpressionAtom // done
+    | base=expressionAtom LEFT_SQUARE_BRACKET index=expressionAtom RIGHT_SQUARE_BRACKET   #subscriptExpression // done
+    | left=expressionAtom bitOperator right=expressionAtom                                #bitExpressionAtom // done
+    | left=expressionAtom mathOperator right=expressionAtom                               #mathExpressionAtom // done
+    | left=expressionAtom comparisonOperator right=expressionAtom                         #binaryComparisonPredicate // done
```
Comment (arnaud-lacurie) [inline fdb-relational-core/src/main/antlr/RelationalParser.g4:1156]: Nice 👍

File: yaml-tests/src/test/resources/arrays.yamsql (hunk 1 above)
Comment (hatyo) [reply yaml-tests/src/test/resources/arrays.yamsql:103]: No, that would require adding support for this operator in the `FunctionKeyExpression` API, so it's not supported out of the box, feel free to open a GitHub issue to implement that feature.

File: yaml-tests/src/test/resources/arrays.yamsql (hunk 2 above)
Comment (hatyo) [reply yaml-tests/src/test/resources/arrays.yamsql:98]: > Could you add a test with X[0]?

done.

> And another test where the array is empty?

we do not have support for empty array literals (see #3588), I added _some_ support to get the tests going, but there are some subtleties around empty array literal's flowing unknown types that requires addressing (see #3589 as a way to explicitly prevent that from happening through explicit casting).

File: fdb-relational-core/src/main/java/com/apple/foundationdb/relational/recordlayer/query/functions/SqlFunctionCatalogImpl.java (hunk 3 above)
Comment (hatyo) [reply fdb-relational-core/src/main/java/com/apple/foundationdb/relational/recordlayer/query/functions/SqlFunctionCatalogImpl.java]: Done.

**Comment by hatyo** [comment]
> Could you call out in the description that the error code has changed for a certain operation? So that we can surface it in our release notes.

If you're referring to this test:

```yaml
      - query: select a, e from ta where e not in ('foo' , 35)
      - initialVersionAtLeast: !current_version
      - error: "42804"
      - initialVersionLessThan: !current_version
      - error: "42601"
```

The difference we see here is purely related to the test itself which I think was poorly written. The test seems to want to verify the SQL error code resulting from having elements in the in-list of incompatible types (`foo` and `35`), however what it was instead verifying a syntax error (`42601`), due to the lack of support of `not in` expression.
With this PR, we _do_ support `not in` so we can verify the data type mismatch (`42804`) which the test intended to verify.

File: yaml-tests/src/test/resources/arrays.yamsql (hunk 1 above)
Comment (arnaud-lacurie) [reply yaml-tests/src/test/resources/arrays.yamsql:103]: https://github.com/FoundationDB/fdb-record-layer/issues/3592
